SIMD-0118: Partitioned Epoch Rewards, amend/extend design (#118)
solana-foundation/solana-improvement-documents: proposals/0118-partitioned-epoch-reward-distribution.md

done as follows.

To minimize the impact on block processing time during the reward distribution
phase, a target of 4,096 stake rewards will be distributed per block. The total
number of blocks `M` needed to distributed rewards is given by the following
formula to round up to the nearest integer without using floating point value
arithmetic:

```
M = ((4096 - 1)+num_stake_accounts)/4096
```

To safeguard against the number of stake accounts growing dramatically and
overflowing the number of blocks in an epoch, the number of blocks is capped at
10% of the total slots in an epoch.

The [SipHash 1-3](https://www.aumasson.jp/siphash/siphash.pdf) pseudo-random
function is used to hash stake account addresses efficiently and uniformly
across the blocks in the reward distribution phase. The hashing function for an
epoch is created by seeding a new `SipHasher` with the parent block's blockhash.
This hashing function can then be used to hash each active stake account's
address into a u64 hash value. The reward distribution block index `I` can then
be computed by applying the following formula to the hash:

```
I = (M * stake_address_hash) / 2^64
```

### Reward Distribution

> The reward distribution phase happens after the reward computation phase, ie.

Reviewer: ie. => i.e.

Author: Given that we have no agreed upon style guide, I'm going to stick with my stylistic preference on this.

   
   num_partitions: u64, // little-endian unsigned 64-bit integer
   
   parent_blockhash: Hash, // byte-array of length 32

   // total rewards for the current epoch, in lamports
   total_rewards: u64, // little-endian unsigned 64-bit integer

   // distributed rewards for the current epoch, in lamports
   distributed_rewards: u64, // little-endian unsigned 64-bit integer
}
```

The `EpochRewards` sysvar is repopulated at the start of the first block of the
epoch (before any transactions are processed), as both the total epoch rewards
and vote account rewards become available at this time. The
`distributed_rewards` field is updated per reward distribution for each block in
the reward distribution phase. The `active` field is set to false after the last
partition is distributed (and `total_rewards == distributed_rewards`), ie.
before processing transactions in block at height `X + M`.

#### Booting from Snapshot

When booting from a snapshot, a node must check the EpochRewards sysvar account
to determine whether the distribution phase is active. If so, the node must
rerun the rewards calculation using the `EpochRewards::num_partitions` and

> `EpochRewards::parent_blockhash` sysvar fields as well as the `EpochStakes` in

Reviewer: `EpochStakes` doesn't store the full stake account state. It only stores the stake delegation info, which is not enough to recompute the rewards. 
To recompute stake rewards, we have to store the full stake account state, which is as expensive as storing the rewards.
Therefore, I think we should keep the current implementation of storing the computed rewards in the snapshot.

Reviewer: > we are guaranteeing that no state will have changed since the first block in the epoch

No, this is not true. When the stake account received rewards, its state changes.

Author: Well, correct. Sorry, I left that part implied. Explain how that is an issue, please.

Reviewer: For example, the `observed_credit` field will change after receiving rewards. 
Reward computation will depend on that field. If we want to recompute rewards later, we need to know the `observed_credit` for all stake accounts at the epoch boundary.

Reviewer: That field is inside stake `state`. but not saved in `EpochStake`. When we load the stake account later, we can't recover the old state.

Author: > that will add extra complexity and make startup from snapshot slower.

The motivation of this amendment is that a little extra time booting from snapshot is preferable to adding 100s of MBs to the snapshot.
I would expect the performance impact of that subtraction to be nominal, but if it is more than that, we definitely don't need to add each entire StakeState to the snapshot to preserve just `credits_observed`.

Reviewer: There is also code complexity and risk aspect to consider. There are a lot of if-else and corner case handling in reward calculation code. If not careful, we could easily introduce a bug. And a bug there will be very bad. 

In my opinion, the extra space in snapshot doesn't sound too bad. 
1. We only pay the cost when the snapshot is taken during reward period. For snaphot take out of reward period, we don't need to pay the cost.
2. I think the reward data should have many similar data in the fields, which will make it very compressible. After compress, it woun't take 100s of MBs.
3. It makes startup from snapshot faster.

Reviewer: I have simulated a bank snapshot with 1M stake rewards, randomly delegated to 2K vote accounts. 

Total space for uncompressed bank snapshot took 296M, compressed space 65M. 

```
-rw-r--r-- 1 sol users 296M Feb 24 00:38 foo.snap
-rw-r--r-- 1 sol users  65M Feb 24 00:38 foo.snap.lz4
```

Reviewer 2: it's only partially about the space. adding snapshot fields must be adopted by all implementation. the fd team was lamenting about this at the offsite and we committed to avoid adding them whenever possible, instead preferring to store data in accounts. storing _this_ data in an account is a nonstarter though.

i don't think we even need to unwind the credits, but we might need to store a couple more fields in the sysvar. we already have the total and distributed lamports. if we also add the total and distributed credits, we can can trivally recover the distributions for the remaining partitions.

Reviewer: I can argue that there is no free lunch. If you don't want to deal with the complexity of handling the new field, you will have to introduce the complexity of a new code path to recompute rewards. And you will have to maintain both reward code paths in future.  Besides, the field is already in v1.17. If fd, use v.1.17 and above as the reference impl, it would already have this field.

Author: Ofc we can ask, but I think it is highly unlikely that firedancer supports partitioned epoch rewards in any fashion yet. The fact that something is already implemented does not make it the right design, especially since the existing approach was not described (approved) in the original simd, nor was that simd signed off on by the firedancer team.

Reviewer: Heh. What's the policy for simd approval, consensus or majority? Anyway, if I need to vote, my vote is against recomputing. But if the majority thinks that recomptuing is the way to go, I am fine to accept it too. [(*^ ‿ ⁢*](https://www.emoticonstext.com/wink.html)

---
simd: '0118'
title: Partitioned Epoch Rewards Distribution
authors: Haoran Yi (Anza), Justin Starry (Anza), Tyera Eulberg (Anza)
category: Standard
type: Core
status: Draft
created: 2024-02-16
feature: (fill in with feature tracking issues once accepted)

> supersedes: '0015'

Author: Certainly something. I like `superceded-by`... will see if that's acceptable to Jacob

> supersedes: '0015'

Author: I also update the simd-0015 header to status: Accepted, because that's technically what it was before this superceding simd. I'm not totally clear on what the status-change process should look like, though (cc @jacobcreech )

> supersedes: '0015'

Reviewer 3: That sounds correct. All of these SIMDs were technically "Accepted" and I'll need to update them. Let me add `superceded-by` so that the CI will play nice. This also helps a lot when we eventually get a website to display SIMDs.

> supersedes: '0015'

Reviewer 3: Added in https://github.com/solana-foundation/solana-improvement-documents/pull/126

function is used to hash stake account addresses efficiently and uniformly
across the blocks in the reward distribution phase. The hashing function for an
epoch is created by seeding a new `SipHasher` with the parent block's blockhash.
This hashing function can then be used to hash each active stake account's
address into a u64 hash value. The reward distribution block index `I` can then
be computed by applying the following formula to the hash:

```
I = (M * stake_address_hash) / 2^64
```

### Reward Distribution

The reward distribution phase happens after the reward computation phase, ie.
during the second block in the epoch as per the current design. It lasts for `M`
blocks (see previous section).

#### `EpochRewards` Sysvar Account

The `EpochRewards` sysvar account records whether the rewards distribution phase
is in progress, as well as the details needed to resume distribution when
starting from a snapshot during the reward distribution phase. These details
include: the parameters needed to recalculate the reward partitions, the total
rewards to be distributed, and the amount of rewards distributed so far.

> The layout of `EpochRewards` sysvar is shown in the following pseudo code.

Reviewer 4: The alignment is somewhat unspecified here. In repr(C), u128 either has 8 or 16 byte alignment depending on architecture. AFAIK bool has 8 byte alignment 🤔

I suggest putting offsets like so `/* 0x08 */` as prefixes to make it unambiguous. 

The padding would obviously be zeros, but it wouldn't hurt to mention here.

Author: We do specify 8-byte alignment in the comment on L152, but certainly doesn't hurt to add offsets, and I'll remove the "ie".
Hopefully that u128 won't be too much of a PITA. If I understand [this issue](https://github.com/rust-lang/rust/issues/54341) correctly, rust will always align u128 to 8 bytes even with repr(C), but C programs may not.

Reviewer 2: move `active` to the end?

Author: Ah, thanks for this, @ripatel-fd . Seems to contradict some of what I read about this.
But anyway, I'll do some reworking and update this on Monday.

Author: Updated to 16-byte alignment, including moving `active` to the end

`EpochRewards::total_rewards` sysvar fields, as well as the `EpochStakes` in the
snapshot. The recalculated rewards can be confirmed by comparing a sum of the
rewards remaining (those partitions expected to not yet have been distributed)
with the difference between the `EpochRewards::total_rewards` and
`EpochRewards::distributed_rewards` fields. Partitions for blocks prior to the
current block height can be discarded.

### Restrict Stake Account Mutation

The stake accounts need to be preserved in credit-only state while distributions
are in progress in order to prevent any changes interfering with reward
distribution.

Therefore, the Stake Program needs access to a syscall which reports whether the
distribution phase is active. This new syscall `sol_get_epoch_rewards_sysvar`
should return the values of the `EpochRewards` sysvar. All Stake Program
instructions that mutate stake data or debit stake balances must be disabled
when `EpochRewards::active` is true.

Any transaction that attempts to invoke such an instruction will fail with this
new error code:

```
StakeError {
   EpochRewardsDistributionActive = 15,
}
```

Other users can access the `active` field from the

> `sol_get_epoch_rewards_sysvar` syscall to determine the distribution-phase

Reviewer 4: Is this syscall going to be specified by another SIMD?

Reviewer 2: we probably need a generic "sysvar getter syscall" simd to cover stuff like this

Author: @ripatel-fd , what do you think of Trent's idea here? There are a lot of these sysvar-getter syscalls already, and it would be nice not to have to define new ones beyond the sysvar layout in SIMDs like this one.

across the blocks in the reward distribution phase. The hashing function for an
epoch is created by seeding a new `SipHasher` with the parent block's blockhash.
This hashing function can then be used to hash each active stake account's
address into a u64 hash value. The reward distribution block index `I` can then
be computed by applying the following formula to the hash:

```
I = (M * stake_address_hash) / 2^64
```

### Reward Distribution

The reward distribution phase happens after the reward computation phase, ie.
during the second block in the epoch as per the current design. It lasts for `M`
blocks (see previous section).

#### `EpochRewards` Sysvar Account

The `EpochRewards` sysvar account records whether the rewards distribution phase
is in progress, as well as the details needed to resume distribution when
starting from a snapshot during the reward distribution phase. These details
include: the parameters needed to recalculate the reward partitions, the total
rewards to be distributed, and the amount of rewards distributed so far.

The layout of `EpochRewards` sysvar is shown in the following pseudo code.

> Prefixes indidate the offset of each field, and padding bytes are always zero.

Reviewer 5: ```suggestion
Prefixes indicate the offset of each field, and padding bytes are always zero.
```

> Prefixes indidate the offset of each field, and padding bytes are always zero.

Reviewer 5: ```suggestion
Prefixes indicate the offset of each field, and padding bytes are always zero.
```

```
#[repr(C)] // C representation, 8-byte alignment
struct EpochRewards{
   // whether the rewards period (calculation and distribution) is active
   // byte. false: 0x00, true: 0x01

>    /* 0x00 */ active: bool,

Reviewer 5: What is the size of `bool`? Can you just be explicit about the size of the type: u8, u16, u32, or u64 please

Reviewer 5: and endianness

   // little-endian unsigned 64-bit integer
   /* 0x08 */ distribution_starting_block_height: u64,
   
   // little-endian unsigned 64-bit integer
   /* 0x10 */ num_partitions: u64,
   
   // byte-array of length 32
   /* 0x18 */ parent_blockhash: Hash,

   // total points calculated for the current epoch, where points equals the sum
   // of delegated stake * credits observed for all delegations
   // little-endian unsigned 128-bit integer
   /* 0x38 */ total_points: u128, 

   // total rewards for the current epoch, in lamports
   // little-endian unsigned 64-bit integer
   /* 0x48 */ total_rewards: u64,

   // distributed rewards for the current epoch, in lamports
   // little-endian unsigned 64-bit integer
   /* 0x50 */ distributed_rewards: u64, 
}
```

> The `EpochRewards` sysvar is repopulated at the start of the first block of the

Reviewer 5: We need to add detail on when the sysvar is updated during the calculation phase and the distribution phase. That is to say, the reward calculation and sysvar update therein should probably be updated before the transactions in the block (or after, i dont think it matters, actually after is better). The rewards distribution phase and sysvar update should be done _after_ the transactions like rent collection.

Reviewer 5: Okay before is fine. I think I could have been clearer that it is preferable to keep it consistent.

partition is distributed (and `total_rewards == distributed_rewards`), ie.
before processing transactions in block at height `X + M`.

#### Booting from Snapshot

When booting from a snapshot, a node must check the EpochRewards sysvar account
to determine whether the distribution phase is active. If so, the node must
rerun the rewards partitioning using the `EpochRewards::num_partitions` and
`EpochRewards::parent_blockhash` sysvar fields and determining the upcoming
partitions by comparing its current block height to
`EpochRewards::distribution_starting_block_height`. Then the runtime must
recalculate the remaining rewards using the `EpochRewards::total_points` and
`EpochRewards::total_rewards` sysvar fields, as well as the `EpochStakes` in the
snapshot. The recalculated rewards can be confirmed by comparing a sum of the
rewards remaining (those partitions expected to not yet have been distributed)
with the difference between the `EpochRewards::total_rewards` and
`EpochRewards::distributed_rewards` fields. Partitions for blocks prior to the
current block height can be discarded.

### Restrict Stake Account Mutation

> The stake accounts need to be preserved in credit-only state while distributions

Reviewer 5: I do not understand the issue here. If the account rewards are credited at the end of the block, then all edits can proceed as normal in txns. Can you elaborate?

Author: In this design, if a node boots from snapshot in the middle of the rewards period, it needs to recalculate the reward partitions. Therefore, Delegations need to remain unchanged until all partitions are distributed, as do the number and addresses of stake accounts.

The original implementation of SIMD-0015 prevented any use of the stake program at all during partitioned distribution. However, given the existing use-cases for depositing into stake accounts (MEV), we wanted to relax this restriction. Credit-only is a conservative approach to doing so.

Certainly merges and splits must be prevented during distribution, and we would have to be very careful about allowing delegate or deactivate. But other operations could be permitted.

That said, I lean toward establishing this SIMD with the conservative credit-only approach and amending the SIMD later as we feel confident about the safety of other operations. Wdyt?

Reviewer 5: That is fine for now. This text should make it into the SIMD

Author: Pushed

Reviewer 5: That looks good

> ---

Author: Hmm, that authors list is particularly not good. I'll see if I can force it to be a list within the table display

> ---

Reviewer 2: yeah and the zero-prefixed numbers being interpreted as octal is... surprising

> ---

Author: > yeah and the zero-prefixed numbers being interpreted as octal is... surprising

Ah, that one I can fix by using single quotes. But agreed.

> ---

Reviewer 2: what does this look like?
```
accounts: Haoran Yi (Anza)<br>Justin Starry (Anza)<br>Tyera Eulberg (Anza)
```

> ---

Reviewer 2: :sob: 

how about
```
accounts: >-
Haoran Yi (Anza)
Justin Starry (Anza)
Tyera Eulberg (Anza)
```

> ---

Reviewer 3: Looks like EIP just uses a [comma separated list](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1559.md). I'd be fine with that since it has a precedent. I'll update the other multi-author SIMDs later if we go with that.

> ---

Author: > how about

It looks like this 😂 
<img width="1067" alt="Screenshot 2024-02-20 at 10 47 55 PM" src="https://github.com/solana-foundation/solana-improvement-documents/assets/3030561/fc523497-51a9-4d7a-82c8-fb9156dc74f5">

I'll push comma-separated list

3. A completely asynchronous epoch rewards calculation and distribution, in
   which both reward computation and rewards distribution are asynchronous.
   This is the most general approach. However, it is also the most complex
   approach. The reward computation states would have to be maintained across
   multiple blocks. The transition between the reward calculation and reward
   distribution would need to be synchronized across the cluster. And cluster
   restart during reward computation would have to be handled specially.

4. An other approach is similar to the current proposal with additional
   per-block reward reserve sysvar accounts. Those sysvar accounts are
   introduced to track and verify the rewards distributed per block. The
   per-block reward reserve sysvar accounts add additional checks and safety
   for reward distribution. However, they also add addition cost to block
   processing, especially for the first block in the epoch. The first block is
   already computationally heavily - it is responsible for processing all the
   reward computation. The additional cost of those sysvars puts more burden
   onto that block and hurt the timing for it.

> ## Detailed Design

Author: I added a couple of my own edits to your suggestions, in the new commit. See what you think.

vote account and also simplifies the overall changes. It also lets us focus on
solving the primary bottleneck - Stake Rewards. Only Stake rewards are going to
be distributed out over multiple blocks.

In the new stake rewards distribution approach, we will separate the
computation of rewards from the actual distribution of rewards at the epoch
boundary by dividing the process into two distinct phases:

   1. rewards calculation phase - during which the epoch rewards for all
      activate stake accounts are computed and distribution chunks are scheduled.

   1. rewards distribution phase - during which the calculated epoch rewards
      for the active stake accounts are distributed.

To help maintain the total capital balance and track/verify the reward
distribution during rewarding phases, a new sysvar account, `EpochRewards`, is
proposed. The `EpochRewards` sysvar holds the balance of the rewards that are
pending for distribution.

### Rewards Calculation

Reward calculation phase computes all the rewards that need to be distributed
for the active stake accounts, and partitions the reward into a number of
chunks for distribution in the next phase.

> Currently, on Solana Mainnet Beta with ~550K active stake accounts, epoch reward
> calculation takes around 10 seconds on average. This will make it impossible to
> perform rewards computation synchronous within one block.
> 
> However, there are quite a few promising optimizations that can cut down the
> reward computation time. An experiment for reward calculation optimization
> (https://github.com/solana-labs/solana/pull/31193) showed that we can cut the
> reward calculation time plus vote reward distribution time to around 1s. This
> makes synchronous reward computation and asynchronous reward distribution a
> feasible approach. We also believe that there is room for more optimization to
> further cut down the above timing.
> 
> Therefore, the following design is based on the above optimization. The reward

Reviewer 2: ```suggestion
The reward
```

> calculation will be performed at the first block of the epoch, block height `X`.

Reviewer 2: throughout we assume that the rewards distribution algorithm is known by the implementer. i'm not sure this is a strong assumption

Author: I assume you mean the rewards *calculation* algorithm?
I don't want to fully define that here, but if there's a good doc to link to, that would be good.

Author: Added this link

Once the full rewards are calculated, the rewards will be partitioned into
distribution chunks stored in the bank, which will then be distributed during
the `reward distribution` phase.

To ensure that each block distributes a subset of the rewards in a
deterministic manner for the current epoch, while also randomizing the
distribution across different epochs, the partitioning of all rewards will be
done as follows.

To minimize the impact on block processing time during the reward distribution
phase, a target of 4,096 stake rewards will be distributed per block. The total
number of blocks `M` needed to distributed rewards is given by the following
formula to round up to the nearest integer without using floating point value
arithmetic:

```
M = ((4096 - 1)+num_stake_accounts)/4096
```

To safeguard against the number of stake accounts growing dramatically and
overflowing the number of blocks in an epoch, the number of blocks is capped at

> 10% of the number of block in an epoch (currently 432,000).

Reviewer 2: ```suggestion
10% of the total slots in an epoch.
```

> 10% of the number of block in an epoch (currently 432,000).

Reviewer 2: we make no provision for how this overflow case is actually handled. iirc the original proposal was to "let the last partition take a long time," but the partitioning algorithm doesn't accommodate.

Author: What's written here does describe the current implementation. "Overflow" is actually handled by letting the number of accounts in a partition exceed 4096 by clamping the number of partitions to `slots_per_epoch/10`:
https://github.com/solana-labs/solana/blob/8ad125d0c0688aaf2b62bb95b535ff988ed7f9ac/runtime/src/bank.rs#L1162-L1165

So each slot will go a little slower, rather than distribution extending for more slots. This is probably okay for now, but we probably ought to consider making `MAX_FACTOR_OF_REWARD_BLOCKS_IN_EPOCH` tunable at some point.

To safeguard against the number of stake accounts growing dramatically and
overflowing the number of blocks in an epoch, the number of blocks is capped at
10% of the number of block in an epoch (currently 432,000).

The [SipHash 1-3](https://www.aumasson.jp/siphash/siphash.pdf) pseudo-random
function is used to hash stake account addresses efficiently and uniformly
across the blocks in the reward distribution phase. The hashing function for an
epoch is created by seeding a new `SipHasher` with the parent block's blockhash.
This hashing function can then be used to hash each active stake account's
address into a u64 hash value. The reward distribution block index `I` can then
be computed by applying the following formula to the hash:

```

> I = (M * stake_address_hash) / 2^64

Reviewer 2: ```suggestion
I = (M * stake_address_hash) / (2^64 - 1)
```

Author: Stet? https://github.com/solana-labs/solana/blob/f122e99c4e737b7cc1c2a697cf81e56eb39ccec7/sdk/src/epoch_rewards_hasher.rs#L42

Reviewer 2: ah right. b- math guy here. i have to redo the calculation scratch every time :sweat_smile:

```

### Reward Distribution

The reward distribution phase happens after the reward computation phase, ie.
during the second block in the epoch as per the current design. It lasts for `M`
blocks (see previous section).

#### `EpochRewards` Sysvar Account

> The `EpochRewards` sysvar account records whether the rewards distribution phase

Reviewer: What's the balance on the sysvar account?

Author: In this SIMD, the balance on the sysvar account is the rent-exempt minimum for the amount of data in the account (which I believe is 0.00134328 SOL).

We should definitely have the conversation about whether to track distributions with the SOL balance of this sysvar here.

My personal opinion is that the existing implementation is a little scary when the balance gets low (below rent-exempt minimum or even below zero). Also, as of this SIMD, the sysvar needs to persist, so its balance can no longer be withdrawn to 0. We could fund the account `sum_distributions` on top of the rent-exempt minimum, but I wrote it up without that because I don't think it gains us much over tracking the amount distributed in the sysvar data.

Reviewer 6: I also agree that it's tricky to manage the sysvar lamport balance properly with all the edge cases. I like this new proposal of just tracking it in the data.

Reviewer: Ok. Sounds good to me. This will also make sysvar cache management simpler too!

One thing to note is that the total capital will *not* stay stable during the epoch. Should any user programs or economics factors that depend on this assumption, they will need to be adjusted.

is in progress, as well as the details needed to resume distribution when
starting from a snapshot during the reward distribution phase. These details
include: the parameters needed to recalculate the reward partitions, the total
rewards to be distributed, and the amount of rewards distributed so far.

The layout of `EpochRewards` sysvar is shown in the following pseudo code.

```
struct EpochRewards{
   // whether rewards distribution is active

>    active: bool, // byte

Reviewer 2: ```suggestion
   active: bool, // byte. false: 0x00, true: 0x01
```

epoch (before any transactions are processed), as both the total epoch rewards
and vote account rewards become available at this time. The
`distributed_rewards` field is updated per reward distribution for each block in
the reward distribution phase. The `active` field is set to false after the last
partition is distributed (and `total_rewards == distributed_rewards`), ie.
before processing transactions in block at height `X + M`.

#### Booting from Snapshot

When booting from a snapshot, a node must check the EpochRewards sysvar account
to determine whether the distribution phase is active. If so, the node must
rerun the rewards calculation using the `EpochRewards::num_partitions` and
`EpochRewards::parent_blockhash` sysvar fields as well as the `EpochStakes` in
the snapshot. Then the runtime must resume rewards distribution from the
partition indicated by comparing its current block height to
`EpochRewards::distribution_starting_block_height`. This can be confirmed by
comparing a partial sum of the rewards calculation (those partitions expected to
have been distributed) with the `EpochRewards::distributed_rewards` field.

> 

Reviewer 2: ```suggestion
Partitions for blocks prior to the current block height can be discarded.

```

### Restrict Stake Account Mutation

The stake accounts need to be preserved in credit-only state while distributions
are in progress in order to prevent any changes interfering with reward
distribution.

Therefore, the Stake Program needs access to a syscall which reports whether the
distribution phase is active. This new syscall `sol_get_epoch_rewards_status`
should report the value of `EpochRewards::active`. All Stake Program

> instructions that mutate stake data or debit stake balances should be disabled

Reviewer 2: ```suggestion
instructions that mutate stake data or debit stake balances must be disabled
```

Any transaction that attempts to invoke such an instruction will fail with this
new error code:

```
StakeError {
   EpochRewardsDistributionActive = 15,
}
```

Other users can access the `active` field from the
`sol_get_epoch_rewards_sysvar` syscall to determine the distribution-phase
status from within the SVM.

## Impact

There are the two main impacts of the changes to stake accounts during the
epoch rewarding phase.

The first impact is that stake accounts will see their rewards being credited a
few blocks later in the epoch than before.

The second impact is that users will only be able to credit their stake accounts
during the epoch reward phase. Any other updates will have to wait until the end
of the phase.

> Nonetheless, the overall amount of time that the user must wait before

Reviewer 2: ```suggestion
Nonetheless, the overall amount of wall time that the user must wait before
```

---
simd: '0118'
title: Partitioned Epoch Rewards Distribution
authors: Haoran Yi (Anza), Justin Starry (Anza), Tyera Eulberg (Anza)
category: Standard
type: Core
status: Draft
created: 2024-02-16
feature: (fill in with feature tracking issues once accepted)
supersedes: '0015'
---

## Summary

A new way to distribute epoch rewards across multiple blocks is proposed to
address the current performance problems associated with epoch reward
distribution in a first block of a new epoch.

## Motivation

The distribution of epoch rewards at the start block of an epoch becomes a
significant bottleneck due to the rising number of stake accounts and voting
nodes on the network.

To address this bottleneck, we propose a new approach for distributing the

> epoch rewards over multiple blocks.

Reviewer 7: Just curious - have we measured that it is the distribution (write-back) versus the calculation that is the bottleneck?

Author: Yes, definitely, although it doesn't seem like we have that data summarized concisely in any one place. It seems to be mostly in various places in the #proj-epoch-boundary-optimization channel on discord. For instance, here's a comment about the calculation time: https://discord.com/channels/428295358100013066/960593861732884520/1096146390037561414 (64ms for 550K stake accounts)
Whereas distribution is more like 5-10s.

   /* 0x10 */ num_partitions: u64,
   
   // byte-array of length 32
   /* 0x18 */ parent_blockhash: Hash,

   // total points calculated for the current epoch, where points equals the sum
   // of delegated stake * credits observed for all delegations
   // little-endian unsigned 128-bit integer
   /* 0x38 */ total_points: u128, 

   // total rewards for the current epoch, in lamports
   // little-endian unsigned 64-bit integer
   /* 0x48 */ total_rewards: u64,

   // distributed rewards for the current epoch, in lamports
   // little-endian unsigned 64-bit integer
   /* 0x50 */ distributed_rewards: u64, 
}
```

The `EpochRewards` sysvar is repopulated at the start of the first block of the
epoch (before any transactions are processed), as both the total epoch rewards
and vote account rewards become available at this time. The
`distributed_rewards` field is updated per reward distribution for each block in
the reward distribution phase (before transaction processing). The `active`
field is set to false after the last partition is distributed (and
`total_rewards == distributed_rewards`), ie. before processing transactions in
block at height `X + M`.

> #### Booting from Snapshot

Reviewer 7: Could we be a bit more verbose about what logic happens in this case? In an ideal world client teams should be able to use SIMDs, without looking at code, to implement new features. Thank-you! 🙏

Reviewer 7: As discussed on Slack, lets update the SIMD with more precise implementation details after a POC has been implemented 😄

When booting from a snapshot, a node must check the EpochRewards sysvar account
to determine whether the distribution phase is active. If so, the node must
rerun the rewards partitioning using the `EpochRewards::num_partitions` and
`EpochRewards::parent_blockhash` sysvar fields and determining the upcoming
partitions by comparing its current block height to
`EpochRewards::distribution_starting_block_height`. Then the runtime must
recalculate the remaining rewards using the `EpochRewards::total_points` and
`EpochRewards::total_rewards` sysvar fields, as well as the `EpochStakes` in the
snapshot. The recalculated rewards can be confirmed by comparing a sum of the
rewards remaining (those partitions expected to not yet have been distributed)
with the difference between the `EpochRewards::total_rewards` and
`EpochRewards::distributed_rewards` fields. Partitions for blocks prior to the

> current block height can be discarded.

Reviewer 7: The rewards calculation assumes that the calculation is done at the epoch boundary (for example, using `VoteState::epoch_credits`). We need to make sure that re-calculating rewards when booting off a snapshot doesn't break the calculation, as this assumption is no longer true.

Reviewer 7: In particular, I think `solana_stake_program::stake_state::calculate_stake_points_and_credits` will have to be tweaked slightly (and possibly some other places), as will anywhere which uses `VoteState::epoch_credits`.

Reviewer 7: No, not in the SIMD, as I don't think this will be fully known until it's implemented. Just wanted to flag 😄

snapshot. The recalculated rewards can be confirmed by comparing a sum of the
rewards remaining (those partitions expected to not yet have been distributed)
with the difference between the `EpochRewards::total_rewards` and
`EpochRewards::distributed_rewards` fields. Partitions for blocks prior to the
current block height can be discarded.

### Restrict Stake Account Mutation

In order to ensure partition recalculations will be the same across the
distribution period, the number and addresses of stake accounts, as well as
stake-delegation amounts, must remain unchanged. There are existing use-cases
for crediting stake accounts at epoch boundaries; so as to not disrupt these
credits, but otherwise take the most conservative approach, account credits will
be the only stake-account changes permitted during distribution.

To limit permissible actions, the Stake Program needs access to a syscall which
reports whether the distribution phase is active. This new syscall
`sol_get_epoch_rewards_sysvar` should return the values of the `EpochRewards`
sysvar. All Stake Program instructions that mutate stake data or debit stake

> balances must be disabled when `EpochRewards::active` is true.

Reviewer 2: nit
```suggestion
account `lamports` balances must be disabled when `EpochRewards::active` is true.
```

I = (M * stake_address_hash) / 2^64
```

### Reward Distribution

The reward distribution phase happens after the reward computation phase, ie.
during the second block in the epoch as per the current design. It lasts for `M`
blocks (see previous section).

#### `EpochRewards` Sysvar Account

The `EpochRewards` sysvar account records whether the rewards distribution phase
is in progress, as well as the details needed to resume distribution when
starting from a snapshot during the reward distribution phase. These details
include: the parameters needed to recalculate the reward partitions, the total
rewards to be distributed, and the amount of rewards distributed so far.

The layout of `EpochRewards` sysvar is shown in the following pseudo code.

```

> struct EpochRewards{

Author: Sure, wouldn't hurt.

> struct EpochRewards{

Reviewer 8: Why is the 16-byte alignment necessary? Memcpy with 16-byte aligned values are not supported in our SBF target. This means, we have compiler error when trying to use EpochRewards.

>    // whether rewards distribution is active

Reviewer 6: Is it also true during rewards calculation?

Author: Technically (at least in the current implementation), rewards calculation completes before the sysvar is populated, although both steps happen in the same block and both complete before any transactions are processed.

I think the existing comment is correct, but I am happy to make more explicit if there's additional information that would be helpful here.

Reviewer 6: Ok makes sense. So is it set to active before the first block of an epoch is processed?

Author: > So is it set to active before the first block of an epoch is processed?

Yep, correct

Reviewer 6: Great and maybe worth one more small clarification that if the last partition of rewards is distributed at the beginning of slot X, it will longer be active in slot X when processing transactions. Just want to avoid an off-by-one issue with other implementations.

Author: Added some clarification. Let me know if it reads okay.

comparing a partial sum of the rewards calculation (those partitions expected to
have been distributed) with the `EpochRewards::distributed_rewards` field.

### Restrict Stake Account Mutation

The stake accounts need to be preserved in credit-only state while distributions
are in progress in order to prevent any changes interfering with reward
distribution.

Therefore, the Stake Program needs access to a syscall which reports whether the
distribution phase is active. This new syscall `sol_get_epoch_rewards_status`

> should report the value of `EpochRewards::active`. All Stake Program

Reviewer 6: Rather than creating a new syscall, should we use the `sol_get_epoch_rewards_sysvar` syscall and check `active` there?

Reviewer 6: I don't have a strong preference either way. I'm leaning towards keeping the full sysvar in the syscall just for consistency with other sysvars. And it's easy enough to add a function for `get_epoch_rewards_status` which wraps that syscall to pull out the `active` field.

Author: Yep, makes sense. I pushed a commit describing the `sol_get_epoch_rewards_sysvar` syscall instead.
